### PR TITLE
Replace fields in APIBinding & APIExport for local overlay

### LINKED ISSUE #368: Add APIExport and APIBinding resources to the repo
Currently the GLBC APIExport and corresponding APIBindng that need to be applied inside the GLBC workspace (root:kuadrant) as part of a GLBC deployment are generated on the fly and applied directly to the cluster using this [script](https://github.com/Kuadrant/kcp-glbc/blob/main/utils/create_glbc_api_export.sh) during an initial [deployment](https://github.com/Kuadrant/kcp-glbc/blob/main/utils/deploy.sh#L235). This means that the APIExport and APIBinding is not represented in the kustomization of the deployment anywhere, and changes to it's spec on updates to KCP, example of this happening [here](https://github.com/Kuadrant/kcp-glbc/pull/380), or changes to the contents of either on the GLBC side (New APIS, additional permission claims required) will not be picked up by existing deployments without manual intervention.

The main reason it's not been moved as yet is that both these resources require additional information from the cluster(kubernetes APIExport identityHash) that needs to be injected into each.

* Add a new apiexports kustomization (config/apiexports/kustomization.yaml). 
* Add a new apiexports kustomzation into the local deployment (config/deploy/local/kcp-glbc/apiexports/kustomization.yaml) When built this should output the glbc apiexport resource and apibinding currently being generated by the [script](https://github.com/Kuadrant/kcp-glbc/blob/main/utils/create_glbc_api_export.sh).
 ``` 
config/deploy/local/kcp-glbc/
├── apiexports
│   ├── kustomization.yaml
│   ├── root-kuadrant-glbc-apibinding.yaml
│   ├── root-kuadrant-glbc-apiexport.yaml
│   └── root-kuadrant-kubernetes-apibinding.yaml
```
* Update the deploy script as required to use this kustomization to generate the APIExport and APIBinding and apply it into the glbc workspace. 

Requires: https://github.com/Kuadrant/kcp-glbc/pull/371

### PR DESCRIPTION
**Pull request by david-martin**
Closes #368

## Description of Changes

* Add base exports/bindings to base kustomize dir with dummy identityHash values
* Add an overlay kustomization that replaces values in the APIBinding/APIExport from an env file

## Release and Testing

* Run `make local-setup` 

## Dependencies

n/a

### CODE REVIEW

File: config/apiexports/root-kuadrant-glbc-apibinding.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+apiVersion: apis.kcp.dev/v1alpha1
+kind: APIBinding
+metadata:
+  annotations: {}
+  name: glbc
+spec:
+  acceptedPermissionClaims:
```
Comment (mikenairn) [inline config/apiexports/root-kuadrant-glbc-apibinding.yaml:7]: This is out of date, see https://github.com/Kuadrant/kcp-glbc/pull/380/files#diff-c4ad9f19fe023b995302ab3e011be0f43f3e64a10580f7dfa377e9bd882a23b7R104

File: config/apiexports/root-kuadrant-glbc-apibinding.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  - group: apps
+    identityHash: DUMMY_HASH
+    resource: deployments
+  - group: networking.k8s.io
+    identityHash: DUMMY_HASH
+    resource: ingresses
+  reference:
+    workspace:
+      exportName: glbc-root-kuadrant
+      path: root:kuadrant
```
Comment (mikenairn) [inline config/apiexports/root-kuadrant-glbc-apibinding.yaml:22]: We will want to be able to in our service deployments create x number of these with different params (exportName and path). Not sure we have done that before with kustomize, do you know how we can create a number of the same resource from the same base file?

File: utils/create_glbc_api_export.sh [hunk 3]
```diff
@@ -175,21 +175,15 @@ kubectl get apibinding kubernetes -o json | jq -e .status.boundResources
...
-kubectl apply view-last-applied apiexport ${name} -o yaml > ${OUTPUT_DIR}/${GLBC_WORKSPACE//:/-}-glbc-apiexport.yaml
+create_glbc_api_export "${GLBC_EXPORT_NAME}" "${coreAPIExportIdentityHash}" ${OUTPUT_DIR}/${GLBC_WORKSPACE//:/-}-glbc-apiexport.yaml
 
 ############################################################
 # Create APIBinding for glbc and core APIs                 #
 ############################################################
 
-create_glbc_api_binding "glbc" "${GLBC_EXPORT_NAME}" "${GLBC_WORKSPACE}" "${coreAPIExportIdentityHash}"
-kubectl apply view-last-applied apibinding ${name} -o yaml > ${OUTPUT_DIR}/${GLBC_WORKSPACE//:/-}-glbc-apibinding.yaml
+create_glbc_api_binding "glbc" "${GLBC_EXPORT_NAME}" "${GLBC_WORKSPACE}" "${coreAPIExportIdentityHash}" ${OUTPUT_DIR}/${GLBC_WORKSPACE//:/-}-glbc-apibinding.yaml
```
Comment (mikenairn) [inline utils/create_glbc_api_export.sh:189]: Is any of this file needed any more I was thinking it would just go away

File: config/deploy/local/kcp-glbc/apiexports/kustomization.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+resources:
+  - ../../../../apiexports
+
+patchesStrategicMerge:
+  - root-kuadrant-glbc-apibinding.yaml
+  - root-kuadrant-glbc-apiexport.yaml
```
Comment (mikenairn) [inline config/deploy/local/kcp-glbc/apiexports/kustomization.yaml:6]: Should these patch files not be committed?

File: config/apiexports/root-kuadrant-glbc-apibinding.yaml (hunk 2 above)
Comment (david-martin) [reply config/apiexports/root-kuadrant-glbc-apibinding.yaml:22]: Using kustomize, we can either generate a patch file, or do some specific patching of fields.
Kustomize doesn't have any templating (any probably won't ever) - https://github.com/kubernetes-sigs/kustomize/issues/4120
My understanding is you either commit files with the actual changes you want in them,
or use something else to do the templating, like a helm chart.

File: utils/create_glbc_api_export.sh (hunk 3 above)
Comment (david-martin) [reply utils/create_glbc_api_export.sh:189]: It's generating the patch files now instead, which are actually just the full APIBinding and APIExport (for simlicity).
So the outcome of calling this script has changed. It outputs files instead of creating them.
We could rename the file & function to reflect this change?

File: config/deploy/local/kcp-glbc/apiexports/kustomization.yaml (hunk 4 above)
Comment (david-martin) [reply config/deploy/local/kcp-glbc/apiexports/kustomization.yaml:6]: These are generated when you run `make local-setup`.
We could check them in, but the identityHash would be different each time you run it, leaving an unclean local repo.

File: utils/create_glbc_api_export.sh (hunk 3 above)
Comment (mikenairn) [reply utils/create_glbc_api_export.sh:193]: right, so we aren't really gaining much from this other than this script no longer applies the kustomization. It will still be a manual process to update each APIExport/APIBinding if the spec or contents change.

File: config/apiexports/root-kuadrant-glbc-apibinding.yaml (hunk 2 above)
Comment (mikenairn) [reply config/apiexports/root-kuadrant-glbc-apibinding.yaml:22]: If we can't use kustomize to do anything useful, is there any point in doing this? Seems like we have created a bunch of kustomize files that we can't actually use in an overlay anyway, no? We are just generating apiexport/apibindings with the create_glbc_api_export.sh that we could just apply directly with `kubectl apply -f blah.yaml`

File: config/deploy/local/kcp-glbc/apiexports/kustomization.yaml (hunk 4 above)
Comment (mikenairn) [reply config/deploy/local/kcp-glbc/apiexports/kustomization.yaml:6]: So not really patches and could just be applied with `kubectl apply -f root-kuadrant-glbc-apibinding.yaml` right?

File: utils/create_glbc_api_export.sh (hunk 3 above)
Comment (david-martin) [reply utils/create_glbc_api_export.sh:193]: Yes, exactly.
Some spec changes to the base files may not need the patch files to be updated though.
We could look at doing this via a json patch where just the identityHashes are overridden? 
https://github.com/kubernetes-sigs/kustomize/blob/master/examples/jsonpatch.md
i.e. change the script to generate the json patch code.

File: config/deploy/local/kcp-glbc/apiexports/kustomization.yaml (hunk 4 above)
Comment (david-martin) [reply config/deploy/local/kcp-glbc/apiexports/kustomization.yaml:6]: Technically yes.
We could strip out anything that shouldn't be different from the base so it's only the identityHash that get's patched,
or we could use a json patch of the identityHash fields?

**Comment by mikenairn** [comment]
@david-martin I had a little dig around and i think we can actually do what we want to do using replacements https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/replacements/. It's a little bit of a faff IMO, but it does work, put an example together here https://github.com/Kuadrant/kcp-glbc/compare/main...mikenairn:kcp-glbc:kustomize_replacement_example?expand=1

If we make the deploy script generate config/deploy/local/kcp-glbc/apiexports/glbc/glbc-apiexport-config.env with the following contents: 
``` 
EXPORT_NAME=glbc-blah
EXPORT_PATH=root:testws
IDENTITY_HASH=thisisatestidentityhashtoinject
```   
Running the kustomize command should produce the desired resources: 
```
$ kustomize build config/deploy/local/kcp-glbc/apiexports/ 
apiVersion: v1
data:
  EXPORT_NAME: glbc-blah
  EXPORT_PATH: root:testws
  IDENTITY_HASH: thisisatestidentityhashtoinject
kind: ConfigMap
metadata:
  name: glbc-apiexport-config-492kkt6h45
---
apiVersion: apis.kcp.dev/v1alpha1
kind: APIBinding
metadata:
  name: glbc
spec:
  permissionClaims:
  - group: ""
    resource: secrets
    state: Accepted
  - group: ""
    identityHash: thisisatestidentityhashtoinject
    resource: services
    state: Accepted
  - group: apps
    identityHash: thisisatestidentityhashtoinject
    resource: deployments
    state: Accepted
  - group: networking.k8s.io
    identityHash: thisisatestidentityhashtoinject
    resource: ingresses
    state: Accepted
  reference:
    workspace:
      exportName: glbc-blah
      path: root:testws
---
apiVersion: apis.kcp.dev/v1alpha1
kind: APIExport
metadata:
  name: glbc-blah
spec:
  latestResourceSchemas:
  - latest.dnsrecords.kuadrant.dev
  - latest.domainverifications.kuadrant.dev
  permissionClaims:
  - group: ""
    resource: secrets
  - group: ""
    identityHash: thisisatestidentityhashtoinject
    resource: services
  - group: apps
    identityHash: thisisatestidentityhashtoinject
    resource: deployments
  - group: networking.k8s.io
    identityHash: thisisatestidentityhashtoinject
    resource: ingresses
 ```

Note also that having the kustomization in it's own directory in the overlay allows us to also create multiple APIExports in an overlay, but use the same base resources from this repo.

wdyt?

**Comment by david-martin** [comment]
@mikenairn That looks OK.
The env file would be generated by the local deploy script,
or be checked in for known environments ahead of time.
I don't think the ConfigMap has any use after the replacement, though it's not likely to cause problems.
Would it make sense to add a `reject` after so it doesn't get applied to the cluster?

**Comment by mikenairn** [comment]
> @mikenairn That looks OK. The env file would be generated by the local deploy script, or be checked in for known environments ahead of time. I don't think the ConfigMap has any use after the replacement, though it's not likely to cause problems. Would it make sense to add a `reject` after so it doesn't get applied to the cluster?

Yeah, it's not required to actually be on the cluster, so if you can stop it applying that, sure, but it's also not going to any harm if it is.

File: utils/deploy.sh [hunk 5]
```diff
@@ -217,15 +199,13 @@ create_sync_target ${GLBC_SYNC_TARGET_NAME}
...
 ${KUSTOMIZE_BIN} build ${KCP_GLBC_KUSTOMIZATION_DIR}/kcp-contrib | kubectl apply -f -
 
 ## Register CertManager APIs with KCP
-${KUSTOMIZE_BIN} build ${CERT_MANAGER_KUSTOMIZATION_DIR}/kcp-contrib | kubectl apply -f -
-
-create_api_binding "cert-manager" "cert-manager-stable" "${GLBC_WORKSPACE}"
+${KUSTOMIZE_BIN} build --reorder none ${CERT_MANAGER_KUSTOMIZATION_DIR}/kcp-contrib | kubectl apply -f -
 
 ############################################################
 # Setup GLBC APIExport                                     #
```
Comment (david-martin) [inline utils/deploy.sh:211]: @mikenairn Would it make sense to only run the `create_glbc_api_export` script when the `DEPLOY_COMPONENTS` is not `none`?
We could remove the `exit 0` and let it fail gracefully if the APIBinding and APIExport can't be created after the sync targets are set up.

File: config/deploy/local/kcp-glbc/apiexports/kustomization.yaml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      - select:
+          kind: APIBinding
+          name: glbc
+        fieldPaths:
+          - spec.reference.workspace.exportName
+      - select:
+          kind: APIExport
+          name: glbc
+        fieldPaths:
+          - metadata.name
```
Comment (mikenairn) [inline config/deploy/local/kcp-glbc/apiexports/kustomization.yaml:53]: Looking ta this again, i think we can move this up into the base of the apiexport `config/apiexports/glbc/kustomization.yaml` and keep have a glbc-apiexport-config.env in that dir with some default values. You can then override it in the overlay like https://github.com/Kuadrant/kcp-glbc/blob/main/config/deploy/local/kcp-glbc/kustomization.yaml#L16-L20

This shoudl avoid a lot of duplication in each of the overalys. wdyt?

File: config/apiexports/kustomization.yaml [hunk 7]
```diff
@@ -0,0 +1,3 @@
+resources:
+  - root-kuadrant-glbc-apibinding.yaml
+  - root-kuadrant-glbc-apiexport.yaml
```
Comment (mikenairn) [inline config/apiexports/kustomization.yaml:3]: nit: I think the references to `root-kuadrant` in the name of these files could be removed now.

File: utils/deploy.sh (hunk 5 above)
Comment (mikenairn) [reply utils/deploy.sh:211]: I suppose it might be better for it to actually fail if you are expecting it to install something and it can't. Having it exit with a non zero exit code if `DEPLOY_COMPONENTS` isn't `none` probably makes sense, can't think of any reason not to anyway.

File: utils/create_glbc_api_export.sh [hunk 8]
```diff
@@ -166,8 +103,9 @@ print_env
 ## Check kube workspace has appropriate kubernetes config   #
 #############################################################
 
-${KUBECTL_KCP_BIN} workspace use ${GLBC_WORKSPACE_KUBERNETES}
+${KUBECTL_KCP_BIN} workspace use ${GLBC_WORKSPACE}
```
Comment (mikenairn) [inline utils/create_glbc_api_export.sh:106]: Why this change?

File: utils/deploy.sh [hunk 9]
```diff
@@ -217,28 +199,41 @@ create_sync_target ${GLBC_SYNC_TARGET_NAME}
...
 
-if OUTPUT_DIR=${KCP_GLBC_KUSTOMIZATION_DIR}/apiexports ${DEPLOY_SCRIPT_DIR}/create_glbc_api_export.sh -w "${GLBC_WORKSPACE}" -W "${GLBC_WORKSPACE}" -n "${GLBC_EXPORT_NAME}"; then
+if OUTPUT_DIR=${KCP_GLBC_KUSTOMIZATION_DIR}/apiexports/glbc ${DEPLOY_SCRIPT_DIR}/create_glbc_api_export.sh -w "${GLBC_WORKSPACE}" -W "${GLBC_WORKSPACE}" -n "${GLBC_EXPORT_NAME}"; then
   echo "GLBC APIExport created successfully for ${GLBC_WORKSPACE} workspace!"
 else
   echo "GLBC APIExport could not be created!"
   # If the GLBC APIExport can't be created, we shouldn't continue to try and deploy anything!
-  exit 0
+  exit 1
 fi
```
Comment (mikenairn) [inline utils/deploy.sh:226]: Is the if statement required now?

File: utils/deploy.sh [hunk 10]
```diff
@@ -217,28 +199,41 @@ create_sync_target ${GLBC_SYNC_TARGET_NAME}
...
 
+## Apply GLBC APIExport
+APIEXPORT_DIR=${KCP_GLBC_KUSTOMIZATION_DIR}/apiexports
+${KUSTOMIZE_BIN} build ${APIEXPORT_DIR} | kubectl apply -f -
+kubectl wait --timeout=120s --for=condition=Ready=true apibinding "glbc"
+kubectl wait --timeout=60s --for=condition=VirtualWorkspaceURLsReady=true apiexport ${GLBC_EXPORT_NAME}
+
+## Dump the APIBinding to a file for use in user workspaces locally
+${KUSTOMIZE_BIN} build ${APIEXPORT_DIR} | yq 'select(.kind == "APIBinding")' > ${APIEXPORT_DIR}/glbc/glbc-apibinding.yaml
```
Comment (mikenairn) [inline utils/deploy.sh:235]: Adding a dependency of `yq` here, which is probably fine, but not one we had before. We do use jq elsewhere, not sure if we list prerequisites anywhere, but if we do, we should add this.

File: utils/local-setup.sh [hunk 11]
```diff
@@ -239,8 +239,8 @@ echo ""
 echo "       cd ${PWD}"
 echo "       export KUBECONFIG=${KUBECONFIG_KCP_ADMIN}"
 echo "       ./bin/kubectl-kcp ws '~'"
-echo "       kubectl apply -f config/deploy/local/kcp-glbc/apiexports/root-kuadrant-kubernetes-apibinding.yaml"
-echo "       kubectl apply -f config/deploy/local/kcp-glbc/apiexports/root-kuadrant-glbc-apibinding.yaml"
+echo "       kubectl apply -f config/apiexports/kubernetes/kubernetes-apibinding.yaml"
+echo "       kubectl apply -f config/deploy/local/kcp-glbc/apiexports/glbc/glbc-apibinding.yaml"
```
Comment (mikenairn) [inline utils/local-setup.sh:243]: @ChristianZaccaria fyi, more path changes. Are there any places in the docs this would need to be changed that aren't already covered in this PR?

File: config/deploy/local/kcp-glbc/apiexports/glbc/kustomization.yaml [hunk 12]
```diff
@@ -0,0 +1,53 @@
+resources:
+  - ../../../../../apiexports
```
Comment (mikenairn) [inline config/deploy/local/kcp-glbc/apiexports/glbc/kustomization.yaml:2]: We need to make change this so there are 2 separate kustomizations, one for the export, and for the binding, This is fine here where we want to apply both, but it makes it impossible to just generate and apply an apiexport without applying the binding in the overlay. We want to be able to use the base here for the APIExport, and APIBindings, but not necessarily apply both of them.

File: utils/create_glbc_api_export.sh (hunk 8 above)
Comment (david-martin) [reply utils/create_glbc_api_export.sh:106]: Remnant of a copy/paste during an iteration where some of this was moved to a separate script

File: utils/deploy.sh (hunk 10 above)
Comment (david-martin) [reply utils/deploy.sh:235]: yq is referenced in sample.sh and tutorial.md
I can swap it out for jq, though kustomize doesn't seem to allow output as json, so it would need another transform step (unless I'm missing something obvious).

Either way, this may be a moot point if I'm understanding https://github.com/Kuadrant/kcp-glbc/pull/386#discussion_r975032354 correctly

Comment (mikenairn) [reply utils/deploy.sh:235]: Yep, wasn't a requirement for install or local setup though i don't think, Either way as you say it doesn't matter now as we will need to do this differently

File: utils/local-setup.sh (hunk 11 above)
Comment (ChristianZaccaria) [reply utils/local-setup.sh:243]: It looks like everything is covered in this PR. The location-api/sample.sh script is the one used in the tutorial, which applies the api bindings. That path was changed in this PR so looks good to me :+1:
